[BUG] Fix forecaster based imputation strategy `Imputer` if forecaster requires `fh` in `fit`

MCRE-BE opened the pull request:
#### Reference Issues/PRs

Fixes #4997

#### What does this implement/fix? Explain your changes.
Passes the missing ForecastingHorizon to the fit method
Adds a test to check with a forecaster requiring fh in fit and not only in predict.

#### Does your contribution introduce a new dependency? If yes, which one?
No

#### Did you add any tests for the change?
Yes

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, I've added myself and possibly others to the [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) file - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

MCRE-BE commented:
@fkiraly : Any advice on the forecaster to use ? Based on docstring alone I can't determine which one to use 😨

fkiraly commented:
You can find all forecasters that don't require `fh` in `fit` as follows

```python
from sktime.registry import all_estimators

all_estimators(filter_tags={"requires-fh-in-fit": False}, as_dataframe=True)
```

It looks like these are all composites - so the `sklearn` reduction forecaster looks like a good choice? `sklearn` is a core dependency, so we do not introduce additional dependencies in the test, from using your example.

fkiraly commented:
So, this one should be ok: `forecaster = make_reduction(GradientBoostingRegressor(), strategy="multioutput")`

For better runtime, perhaps replace with linear regressio?n

MCRE-BE commented:
@fkiraly : Done. Test is added.
In theory, should be ok for testing.

fkiraly commented:
Linting is failing - the long version on how to set it up locally is here: https://www.sktime.net/en/latest/developer_guide/coding_standards.html

You can also click on "details" (at the right of this:
![image](https://github.com/sktime/sktime/assets/7985502/e6f3cb69-fcd7-472c-a735-67b956a036c3)
) to see what's wrong

fkiraly reviewed (approved):
That was quick :-)

looks good now!

MCRE-BE commented:
> Linting is failing - the long version on how to set it up locally is here: https://www.sktime.net/en/latest/developer_guide/coding_standards.html
> 
> You can also click on "details" (at the right of this: ![image](https://user-images.githubusercontent.com/7985502/257555616-e6f3cb69-fcd7-472c-a735-67b956a036c3.png) ) to see what's wrong

Thanks. I saw. I added a whitespace after the docstring. Fixed.

I didn't bother installing a new environment just for this small fix. So I don't have all the same linting. Sorry.

fkiraly commented:
>I didn't bother installing a new environment just for this small fix.

No worries, that's understandable.

fkiraly commented on .all-contributorsrc:
```diff
@@ -2276,6 +2276,15 @@
+      "avatar_url": "https://avatars.githubusercontent.com/u/99316631",
+      "profile": "https://github.com/MCRE-BE",
+      "contributions": [
+        "bug"
```
you can also get the `code` badge, because you also fixed the bug

MCRE-BE commented:
I'll add it tomorrow. PC is turned off 😜

MCRE-BE commented:
Done.

MCRE-BE commented:
It seems all fails are because of urllib...
